internal/pipeline/parser_test.go

```diff
@@ -1488,3 +1488,438 @@ steps:
+}`,
+		},
+		{
+			desc: "Matrix with multiple dimensions, adjustments, and interpolation",
+			input: `---
+env:
+  OS_KEY: os
+  FAVOURITE_OS: darwin
+  FAVOURITE_ARCH: arm64
+  BIG_IRON_OS: zos
+  BIG_IRON_ARCH: s390x
+steps:
+  - command: GOOS={{matrix.os}} GOARCH={{matrix.arch}} go build
+    matrix:
+      setup:
+        "${OS_KEY}":
+          - windows
+          - linux
+          - ${FAVOURITE_OS}
+        arch:
+          - ${FAVOURITE_ARCH}
+          - amd64
+      adjustments:
+        - with:
+            "${OS_KEY}": ${BIG_IRON_OS}
+            arch: ${BIG_IRON_ARCH}
+          soft_fail: true
+`,
+			want: &Pipeline{
+				Env: ordered.MapFromItems(
+					ordered.TupleSS{Key: "OS_KEY", Value: "os"},
+					ordered.TupleSS{Key: "FAVOURITE_OS", Value: "darwin"},
+					ordered.TupleSS{Key: "FAVOURITE_ARCH", Value: "arm64"},
+					ordered.TupleSS{Key: "BIG_IRON_OS", Value: "zos"},
+					ordered.TupleSS{Key: "BIG_IRON_ARCH", Value: "s390x"},
+				),
+				Steps: Steps{
+					&CommandStep{
+						Command: "GOOS={{matrix.os}} GOARCH={{matrix.arch}} go build",
+						Matrix: &Matrix{
+							Setup: MatrixSetup{
+								"arch": {"arm64", "amd64"},
+								"os":   {"windows", "linux", "darwin"},
+							},
+							Adjustments: MatrixAdjustments{
+								{
+									With: MatrixAdjustmentWith{
+										"arch": "s390x",
+										"os":   "zos",
+									},
+									SoftFail: true,
+								},
+							},
+						},
+					},
+				},
+			},
+			wantJSON: `{
+  "env": {
+    "OS_KEY": "os",
+    "FAVOURITE_OS": "darwin",
+    "FAVOURITE_ARCH": "arm64",
+    "BIG_IRON_OS": "zos",
+    "BIG_IRON_ARCH": "s390x"
+  },
+  "steps": [
+    {
+      "command": "GOOS={{matrix.os}} GOARCH={{matrix.arch}} go build",
+      "matrix": {
+        "adjustments": [
+          {
+            "soft_fail": true,
+            "with": {
+              "arch": "s390x",
+              "os": "zos"
+            }
+          }
+        ],
+        "setup": {
+          "arch": [
+            "arm64",
+            "amd64"
+          ],
+          "os": [
+            "windows",
+            "linux",
+            "darwin"
+          ]
+        }
+      }
+    }
+  ]
+}`,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			//t.Parallel()
+			got, err := Parse(strings.NewReader(test.input))
+			if err != nil {
+				t.Fatalf("Parse(%q) error = %v", test.input, err)
+			}
+			if err := got.Interpolate(nil); err != nil {
+				t.Fatalf("Pipeline.Interpolate(nil) = %v", err)
+			}
+			if diff := cmp.Diff(got, test.want, cmp.Comparer(ordered.EqualSS)); diff != "" {
+				t.Errorf("parsed pipeline diff (-got +want):\n%s", diff)
+			}
+			gotJSON, err := json.MarshalIndent(got, "", "  ")
+			if err != nil {
+				t.Fatalf(`json.MarshalIndent(got, "", "  ") error = %v`, err)
+			}
+			if diff := cmp.Diff(string(gotJSON), test.wantJSON); diff != "" {
+				t.Errorf("marshalled JSON diff (-got +want):\n%s", diff)
+			}
+		})
+	}
+}
```

Comment:
this file is getting really long - would it make sense to split it up a bit for easier navigation? happy for this to be done as part of a separate PR

Reply:
I've moved the new tests, but the rest can remain for now.